### DISCUSSION TITLE
Feedback/discussion for #459: Building Docs from Source-Based Libraries

### DISCUSSION THREAD

**Discussion by lread** [cljdoc/cljdoc#460] [Ideas]
This is where we'll figure out our first cut of having cljdoc support importing docs from public git repos.
The results of these discussions will be distilled in #459, which will act as a design doc.

**Comment by lread** [comment]
Topic: How cljdoc learns about public git repos

**Reply by lread** [reply]
Transcribed from git issue comments:

@seancorfield wrote:
> I don't know how often cljdoc checks clojars but whenever I make a new release of a lib (to clojars) and then I go to cljdoc, it always asks me if I want to build a new set of docs (i.e., it hasn't already figured out a new release exists and imported it for me).

@lread responded:
> Not sure if you are aware, but there is a [REST API to trigger an import of docs from clojars](https://github.com/cljdoc/cljdoc/blob/c12eaae2f40447bc1009af55fbe41566f6d152f3/doc/docsets.md#trigger-a-build-for-a-project), [here's how I use it for rewrite-clj](https://github.com/clj-commons/rewrite-clj/blob/f7d0f035b090650bdee3be7f4bdd5c1e70beb338/script/ci_release.clj#L149-L154).

@lread also responded:
> If I've understood correctly, [cljdoc checks clojars every minute for new releases and then queues up ingest (aka doc import) jobs for any new releases every 10 minutes](https://github.com/cljdoc/cljdoc/blob/c12eaae2f40447bc1009af55fbe41566f6d152f3/src/cljdoc/server/release_monitor.clj#L94-L95).

**Comment by lread** [comment]
Topic: Cljdoc rules for accepting/rejecting a git hosted repo

**Reply by lread** [reply]
Transcribed from git issue comments:

**Reply by lread** [reply]
If I have a `cljdoc.edn`, obviously I intend to have a version of my docs hosted on cljdoc.  So that's a given.

For a source-based lib to be imported it will need to have a git version tag.  
Wondering out loud: what's to stop someone from importing docs from my source-lib from an arbitrary git tag I've happened to create for my in my repo? Do I care? Should cljdoc only import from tags that look like versions?

**Reply by seancorfield** [reply]
What does a version look like? People use a lot of different version formats in the wild. Might you want to restrict it to versions that start with a digit? Maybe that's reasonable to prevent arbitrary tags/branches being imported.

A tag is going to either be just the version or `v` followed by the version I think? And then the additional tag would be `cljdoc/` + version?

**Reply by lread** [reply]
> A tag is going to either be just the version or `v` followed by the version I think? And then the additional tag would be `cljdoc/` + version?

Yes, that's what we'll expect. Thinking out loud again: Cljdoc's expectations for version tags (I like your matching scheme) will be documented.  So, a repo owner would know, if they read the docs, that any tags matching `#"^v?\d.*"` (and same prefixed with `cljdoc/`) are fair game for doc import.

**Comment by lread** [comment]
Topic: Where cljdoc looks for articles/source in a git hosted repo

**Reply by lread** [reply]
Transcribed from git issue comments:

@seancorfield wrote:
> `tools.deps.alpha` has specific logic for deducing repo URL from lib name (I posted a link to that specific code in Slack). I don't know if it's public code that cljdoc could use directly but it would be nice to have it follow the same rules as t.d.a.

**Reply by lread** [reply]
@IGJoshua, just thinking out loud here: If clojure tools deps `deps.edn` `:deps` does not give the ability to bring in different sources based on  aliases or alternate paths, I'm wondering if cljdoc needs to.  

We might need to support something akin to its `:deps/root`, but beyond that, I'm not sure... but am happy to learn about something I am missing. Actually, I'm thinking in general folks creating a source-based lib won't be putting their work under a sub-folder so we can probably get away without `:deps/root`-like support (until we learn we actually need it).

Whatcha think?

**Reply by seancorfield** [reply]
I would say don't worry about this initially and see what real-world cases come up. I suspect you will get requests to support `:deps/root` via some machinery _at some point_ (since there are enough repos out there that have multiple projects in them that _someone_ is bound to want docs from source). But that would mean the UI would need to be updated to provide a root folder name field, at least for source-based imports.

**Reply by lread** [reply]
> I would say don't worry about this initially

I like that idea and will run with it.

**Comment by lread** [comment]
Topic: Are there any conflicts between docs imported for lib x version y from Clojars vs same hosted on a GitHub repo?

**Reply by lread** [reply]
Transcribed from git issue comments:

**Reply by lread** [reply]
I suppose there is a slight chance for a race condition:
Alice is getting ready to release on clojars and tags her git repo with a version.
Before she publishes to clojars, Bill notices her repo and triggers a source-based import on cljdoc.

**Reply by seancorfield** [reply]
In that case, Alice could go find the invisible button and re-import it -- and that second time it would pull from Clojars and use the JAR-based workflow?

**Reply by lread** [reply]
Oh right... that IS what you proposed originally isn't it? I think that makes sense, thanks!

Perhaps slightly tangential but also of interest: source-based libs are potentially less permanent than clojars based ones. 
- What if someone imports docs from, then later deletes their repo.  I guess the cljdoc docs live on but links will 404 to the repo. And then docs are not re-importable.
- Or after import, renames or deletes the version tag. Same idea, I guess?

And we have not yet answered your query:
> The complication would be that if an author publishes to Clojars and then decides to do GH-only releases, cljdoc would need some way to know/be told that GH needs to be checked for a new version being requested?

Maybe ok-ish here? If version x of lib y does not exist on clojars than fair game for import from source-based?

Cljdoc automatically detects new versions of libs already imported from clojars.
But... Cljdoc will have to be told by the library author when a new version of their source-based lib is available. 

When a library author switches from clojars to source-based for a new version of their lib, I suppose that auto-checking for new versions on clojars is harmless, because there won't be any new versions from clojars.  

But what tells cljdoc that an import from source-based is the same lib as an import from clojars?  Some sort of absolute canonical scm url?

**Reply by seancorfield** [reply]
> What if someone imports docs from, then later deletes their repo. I guess the cljdoc docs live on but links will 404 to the repo. And then docs are not re-importable.

Same with a Clojars import tho' -- if the source repo goes away later, the github link will be broken, as will any explicit embedded links out to the repo. I link back to the repo from some of my docs so folks can easily click through to see recent changes etc. _[I'm also currently doing this to get around the inability to import docs from a commit after a release, when the docs need to refer to the SHA of the release!]_

For the Clojars -> GH-only issue, yeah, that sounds reasonable.

For that last Q: just the lib name (group/artifact ID). Same as how it would identify any other lib?

**Reply by lread** [reply]
> Same with a Clojars import tho' -- if the source repo goes away later, the github link will be broken, as will any explicit embedded links out to the repo

I guess a minor difference is that a reimport from clojars published lib can be reattempted, the API will reimport from the jar, but any articles will fail to import and links back to scm will be broken.  A reimport attempt from a source-based lib will just outright fail.

>> But what tells cljdoc that an import from source-based is the same lib as an import from clojars? Some sort of absolute canonical scm url?   
>
> For that last Q: just the lib name (group/artifact ID). Same as how it would identify any other lib?

Oh right... I was thinking that group/artifactId might change when moving from clojars to a source-based lib, but that's not true, is it?

**Reply by seancorfield** [reply]
> the API will reimport from the jar, but any articles will fail to import and links back to scm will be broken. A reimport attempt from a source-based lib will just outright fail.

Ah, good point. But then if a source-based lib's repo is deleted, no versions -- past or present -- can be used by anyone so it probably doesn't matter that the cljdoc version also goes away? (unless they've got that library version locally already, in which case they have the local source of the docs too)

**Reply by lread** [reply]
We could call these cljdoc imports orphaned.  
And for the initial release, we can decide to not address them.

**Reply by lread** [reply]
Idle thought: I'm imagining one use case that might occur after source-based libs support is live, is that some folks who publish to clojars might try to use it as a kind of preview before publishing to clojars.  Although maybe not much, as our `cljdoc/<version>` tag + reimport support will make doc corrections after release possible.

**Comment by lread** [comment]
Topic: Immutability

**Reply by lread** [reply]
Transcribed from git issue comments:

**Reply by lread** [reply]
Sounds like we are all good on allowing a moving cljdoc/version tag.

**Comment by lread** [comment]
Topic: Security

**Reply by lread** [reply]
Transcribed from git issue comments:

**Reply by lread** [reply]
I was just thinking here about my idea of perhaps checking inside `cljdoc.edn` for extra optin - and that's probably not something to do absolutely blindly. If we think we want/need this would at least first check if file size is reasonable.

**Reply by lread** [reply]
I think @SevereOverfl0w is security minded? I wonder if he would have time to give #459 all a look-over from that perspective. 🤔

**Comment by seancorfield** [comment]
With the release of Clojure CLI 1.10.3.933, I suspect we're going to see a rise in GH-only releases, at least of tooling. `tools.build` joins the ranks of "common" tooling that has only GitHub releases (they currently auto-gen API docs into that repo but it's pretty horrible). Cognitect's `test-runner` is arguably "very common" for `deps.edn` projects at this point and has no Clojars release (although its API is minimal and therefore its need for cljdoc is also minimal).

I'm still making Clojars releases of `depstar` and `clj-new` but both are "tools"-compatible at this point and can be depended on via just `:git/tag` and a short SHA so I don't _need_ Clojars releases for them any more. I just released `deps-new` (a new, simpler alternative to `clj-new`) that will be GH-only.

It looks like most of the _design_ is sketched out at this point although some of the finer details are missing (e.g., the `cljdoc.edn` flag to opt-in to GH-only publishing; the various API additions). What's the current thinking on this topic?

**Reply by lread** [reply]
It seems like the broad strokes are in, doesn't it? 
I'll  see if I can convert any open questions in #459 to proposals.
I could also start to think about/sketch APIs and technical impacts and see if that triggers any more questions.  

I'd also want to switch primary terms to any Clojure core team terminology where that exists. For example, I think what I have been calling "Public Git Repos" they call "source-based libs".

I think the brunt of the changes will be in the cljdoc-analyzer project, but I'll have to have a thinky-poo.

Hey, did I just volunteer to implement this?

**Reply by lread** [reply]
Just rambling observations:
Your mention of minimal API triggered me to think about libs called from clojure vs clojure tools.
For clojure tools, I'll group in libs invoked from clojure cli and deps.edn. 

For a tool, the the API, I think, might be typically described in the articles. Examples are [test-runner](https://github.com/cognitect-labs/test-runner), [deps-new](https://github.com/seancorfield/deps-new), and my own little effort, [test-doc-blocks](https://github.com/lread/test-doc-blocks).

The docstrings for tools libs will typically refer back to the articles.  
One might argue that clojure API docs aren't even necessary for these projects.  
That might be true, but I think API docs for these projects do add some value as they do describe supported entry points should someone want to call them from clojure code.

And this then makes me think about truly documentation only (or api-less?) projects like [clj-graal-docs](https://github.com/lread/clj-graal-docs).
Adding clojars-less import might attract the interest of documentation-only repos.
I leave this out-of-scope for our initial release.

**Reply by seancorfield** [reply]
`deps-new` has almost no API and it's implementation is marked `^:no-doc` deliberately. All of the value will be in articles. So, yes, it's particularly true that nice docs on cljdoc.org are appealing and, right now, unobtainable without these changes :)

**Comment by lread** [comment]
Testing

**Reply by lread** [reply]
From @IGJoshua on slack:

> Whenever we're ready for testing some of this stuff, my https://github.com/IGJoshua/americano repo has a tag for cljdoc/v1.2.0, and I'd love to get that to be the tag that's built on cljdoc.

**Comment by lread** [comment]
Topic: Search

**Reply by lread** [reply]
I'm just finishing up a maintenance sweep of cljdoc.  
One thing we did not discuss yet (I think?) is how this feature will integrate with search.

**Reply by lread** [reply]
The current search feature is backed by a clojars downloaded database (and some select maven central entries too).
There are plans #308 and works #359 to make search results sort consider popularity. 

Adding some weighting makes a lot of sense.  For example, I would expect to see `com.github.seancorfield/honeysql` appear near the top on a search for `honeysql`, but it currently appears way, way, way down the list.

For clojars, we have the download count. Clojars itself considers the download count, to good effect, in its search result ordering.

But for GitHub (and other SCMs) I'm not sure what we have unrestricted access to.
- Number of stars?
- Number of clones?
- Number of org followers?

Cljdoc does not currently collect any data on browsed docs, but I suppose we could start doing that and feed that back into search result relevance. Other ideas are most welcome.
